app/scripts/keyring-controller.js

```diff
@@ -0,0 +1,572 @@
+  }
+
+  createNewVaultAndKeychain (password, entropy, cb) {
+    this.createNewVault(password, entropy, (err) => {
+      if (err) return cb(err)
+      this.createFirstKeyTree(password, cb)
+    })
+  }
+
+  createNewVaultAndRestore (password, seed, cb) {
+    if (typeof password !== 'string') {
+      return cb('Password must be text.')
+    }
+
+    if (!bip39.validateMnemonic(seed)) {
+      return cb('Seed phrase is invalid.')
+    }
+
+    this.clearKeyrings()
+
+    this.createNewVault(password, '', (err) => {
+      if (err) return cb(err)
+      this.addNewKeyring('HD Key Tree', {
+        mnemonic: seed,
+        n: 1,
+      }, (err) => {
+        if (err) return cb(err)
+        const firstKeyring = this.keyrings[0]
+        const accounts = firstKeyring.getAccounts()
+        const firstAccount = accounts[0]
+        const hexAccount = normalize(firstAccount)
+        this.configManager.setSelectedAccount(hexAccount)
+        this.setupAccounts(accounts)
+
+        this.emit('update')
+        cb(null, this.getState())
+      })
+    })
+  }
+
+  migrateAndGetKey (password) {
+    let key
+    const shouldMigrate = !!this.configManager.getWallet() && !this.configManager.getVault()
+
+    return this.loadKey(password)
+    .then((derivedKey) => {
+      key = derivedKey
+      this.key = key
+      return this.idStoreMigrator.oldSeedForPassword(password)
+    })
+    .then((serialized) => {
+      if (serialized && shouldMigrate) {
+        const keyring = this.restoreKeyring(serialized)
+        this.keyrings.push(keyring)
+        this.configManager.setSelectedAccount(keyring.getAccounts()[0])
+      }
+      return key
+    })
+  }
+
+  createNewVault (password, entropy, cb) {
+    const configManager = this.configManager
+    const salt = this.encryptor.generateSalt()
+    configManager.setSalt(salt)
+
+    return this.migrateAndGetKey(password)
+    .then(() => {
+      return this.persistAllKeyrings()
+    })
+    .then(() => {
+      cb(null)
+    })
+    .catch((err) => {
+      cb(err)
+    })
+  }
+
+  createFirstKeyTree (password, cb) {
+    this.clearKeyrings()
+    this.addNewKeyring('HD Key Tree', {n: 1}, (err) => {
+      const firstKeyring = this.keyrings[0]
+      const accounts = firstKeyring.getAccounts()
+      const firstAccount = accounts[0]
+      const hexAccount = normalize(firstAccount)
+      const seedWords = firstKeyring.serialize().mnemonic
+      this.configManager.setSelectedAccount(firstAccount)
+      this.configManager.setSeedWords(seedWords)
+      autoFaucet(hexAccount)
+      this.setupAccounts(accounts)
+      this.persistAllKeyrings()
+      .then(() => {
+        cb(err, this.getState())
+      })
+      .catch((reason) => {
+        cb(reason)
+      })
+    })
+  }
+
+  placeSeedWords (cb) {
+    const firstKeyring = this.keyrings[0]
+    const seedWords = firstKeyring.serialize().mnemonic
+    this.configManager.setSeedWords(seedWords)
+  }
+
+  submitPassword (password, cb) {
```

Comment:
used in both: unlocking metamask, showing seed words

calls `migrateAndGetKey` - overly close couppling of migration and controller